pallets/kilt-launch/src/lib.rs

```diff
@@ -0,0 +1,797 @@
+
+				// Set locks in target and remove/update storage entries for source
+				Ok(Some(Self::migrate_kilt_balance_lock(&source, &target, Some(amount))?).into())
+			} else {
+				Err(Error::<T>::BalanceLockNotFound.into())
+			}
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Remove KILT balance locks for the specified block
+	fn unlock_balance(block: T::BlockNumber) -> u32 {
+		if let Some(unlocking_balance) = <UnlockingAt<T>>::take(block) {
+			// Remove locks for all accounts
+			for account in unlocking_balance.iter() {
+				<pallet_balances::Pallet<T>>::remove_lock(KILT_LAUNCH_ID, account);
+				<BalanceLocks<T>>::remove(account);
+			}
+
+			Self::deposit_event(Event::Unlocked(block, unlocking_balance.len() as u64));
+			// Safe because `UnlockingAt` will be ~6 in our case
+			unlocking_balance.len() as u32
+		} else {
+			0
+		}
+	}
+
+	/// Transfers all balance of the source to the target address and sets up
+	/// vesting or the KILT balance lock if any of the two were set up
+	/// for the source address.
+	///
+	/// Note: Expects the source address to be an unowned address which was set
+	/// up in the genesis block via `GenesisBuild` and should be claimed by a
+	/// user to migrate to their account.
+	fn migrate_user(source: &T::AccountId, target: &T::AccountId) -> Result<Weight, DispatchError> {
+		// There should be no locks for the source address
+		ensure!(Locks::<T>::get(source).len().is_zero(), Error::<T>::UnexpectedLocks);
+
+		// Transfer to target address
+		let amount = <pallet_balances::Pallet<T>>::total_balance(source);
+		<pallet_balances::Pallet<T> as Currency<T::AccountId>>::transfer(source, target, amount, AllowDeath)?;
+
+		// Migrate vesting info and set the corresponding vesting lock if necessary
+		let mut post_weight: Weight = Self::migrate_vesting(source, target)?;
+
+		// Set the KILT custom lock if necessary
+		post_weight += Self::migrate_kilt_balance_lock(source, target, None)?;
+
+		Ok(post_weight)
+	}
+
+	/// Migrate the vesting schedule from one account to another, if it was set
+	/// in the genesis block via `GenesisBuild`, and set the corresponding
+	/// vesting lock.
+	///
+	/// We already unlock all available funds between the starting and the
+	/// current block. This enables the user to be able to pay for transactions.
+	/// One of these would be `pallet_vesting::vest()` which has to be called
+	/// actively to unlock more of the vested funds.
+	fn migrate_vesting(source: &T::AccountId, target: &T::AccountId) -> Result<Weight, DispatchError> {
+		if let Some(source_vesting) = Vesting::<T>::take(source) {
+			// Check for an already existing vesting schedule for the target account
+			// which would be the case if the claimer requests migration from multiple
+			// source accounts to the same target
+			let vesting = if let Some(VestingInfo {
+				locked,
+				per_block,
+				starting_block,
+			}) = Vesting::<T>::take(&target)
```

Comment:
Maybe that's more readable? Not sure... 🤷🏻 

```rust
let vesting = Vesting::<T>::take(&target).map_or(Ok(source_vesting), |VestingInfo {
		locked,
		per_block,
		starting_block,
	}| {
		// Should never throw because all source accounts start vesting in genesis block
		ensure!(
			starting_block == source_vesting.starting_block,
			Error::<T>::ConflictingVestingStarts
		);
		Ok(VestingInfo {
			// We can simply sum `locked` and `per_block` because of the above requirement
			locked: locked.saturating_add(source_vesting.locked),
			per_block: per_block.saturating_add(source_vesting.per_block),
			starting_block,
		})
	})?;

Reply:
As discussed, the `ensure!` breaks it kinda. I hope this will suffice: d35cf33638fbdd99eca025fc03f817093b2e438e